### PR TITLE
Lazy dtype update (again)

### PR DESCRIPTION
**Pull request by rhattersley**
Another episode in the continuing saga of #1287 (previous episodes: #1290, #1291, and #1502).

### CODE REVIEW

**Comment by rhattersley** [comment]
@ocefpaf & @esc24 - I'd love to get your feedback.

**Comment by rsignell-usgs** [comment]
@ocefpaf, are you going to test this, or do you want me to?

**Comment by ocefpaf** [comment]
@rsignell-usgs I am testing this now, but feel free to run your notebooks and complement the tests.

**Comment by ocefpaf** [comment]
@rhattersley First I apologize for the mess!  It is probably my messy PR #1502 that screwed up the changes.

This PR is OK for most datasets, but I am still getting a `MemoryError` when using it with some `derived_coords`.  This is probably also my fault!  I believe that the derived coords that fail are from my PR #1304.  I made a failed attempt to fix this here in #1533.

Here is a notebook showing the issue:
http://nbviewer.ipython.org/gist/ocefpaf/0833d148d48741446980

I will try to re-visit #1533.  (Unless you have a quick fix for that.)


PS:  PR #1304 started before #1291 (the first chapter of this saga) and shows the importance of a re-factor in the AuxCoord stuff.  As soon as I get some free time I will play with that, even though my `iris` code base knowledge is still limited to make something useful.

**Comment by rhattersley** [comment]
OK, so here's where I think we've got to.

1. We've dodged the issue of cube-merge triggering the load of auxiliary coordinates. iris.load() now works just fine with these large datasets.
2. Printing the cube creates the derived coordinate which triggers the load of its coordinate dependencies. (NB. It's _not_ trying to compute the derived coordinate's values.) In @ocefpaf's example notebook the sea_surface_height_above_geoid is one of dependencies of the ocean sigma aux factory. But sea_surface_height_above_geoid has the shape (48276, 51391) which equates to 9 GB.... hence the MemoryError.

So to avoid (2) we need a way to create derived coordinates that doesn't load its dependencies.

**Comment by rhattersley** [comment]
By extending biggus to allow indexing with np.newaxis and automatic conversion to numpy arrays (see SciTools/biggus#116) I've hacked together a proof-of-concept in Iris which doesn't load derived coordinate dependencies under some common circumstances. Fortunately these circumstances include @ocefpaf's notebook. :wink:

To be more specific:

1. The dimension order of the auxiliary coordinates needs to match the dimension order of the data variable.
2. I think the derived coordinate needs to span all the dimensions of the cube.

Fixing (1) would require adding transpose to biggus.
Fixing (2) should be pretty simple and won't need biggus changes.

**Comment by ocefpaf** [comment]
It works just fine.  I am anxious to start using this to slice my cubes in the vertical. We really need this to compare models and [gliders](http://en.wikipedia.org/wiki/Underwater_glider).

http://nbviewer.ipython.org/gist/ocefpaf/5460e74974c5bd2f6dc2

Travis is not happy, but I am :joy_cat:!  Thanks!!

**Comment by rsignell-usgs** [comment]
Yes, I'm hoping this can be merged soon, so our users can finally stop using our hacked Iris!

**Comment by bjlittle** [comment]
@rhattersley Did you want me to take this PR over and push it through to master so that it makes the cut for the v1.8.0 release?

**Comment by rhattersley** [comment]
> Did you want me to take this PR over and push it through to master so that it makes the cut for the v1.8.0 release?

What did you have in mind?

I'd like to see this in 1.8 if possible, so when I get time (hopefully either tomorrow or the day after) I plan to tackle the existing biggus issue first. That said, I'm very happy for you to move it on quicker than that!

**Comment by rhattersley** [comment]
> Fixing (1) would require adding transpose to biggus.

Which @pelson has kindly done in SciTools/biggus#117. :smile:

**Comment by rhattersley** [comment]
Progress update:
Finishing off the extensions to biggus is proving more awkward than I'd hoped. I'm still working at it, but if I can't get it finished in the next few days then it'll probably be too late to enable this PR to make the v1.8 release. That said, I'm going to push for a quarterly release cycle for Iris, so it shouldn't be too much longer to wait.

**Comment by ocefpaf** [comment]
There is no emoji for sad panda!  But I am following the Biggus developments and it is looking awesome!

So don't hush on our account.  We'll just start building our conda package from your branch again :stuck_out_tongue_winking_eye:

**Comment by rhattersley** [comment]
We have a tick! \o/

**Comment by bjlittle** [comment]
Oh my ... :smile:

**Comment by bjlittle** [comment]
@rhattersley With a fresh pair of :eyes: ... I'm on this right now ...

File: lib/iris/aux_factory.py [hunk 1]
```diff
@@ -279,15 +288,22 @@ def _nd_points(self, coord, dims, ndim):
...
-        # Figure out the n-dimensional shape.
-        nd_shape = [1] * ndim
-        for dim, size in zip(dims, coord.shape):
-            nd_shape[dim] = size
-        points.shape = tuple(nd_shape)
+        # Expand dimensionality to be consistent with the Cube.
+        if dims:
+            keys = [None] * ndim
+            for dim, size in zip(dims, coord.shape):
+                keys[dim] = slice(None)
```
Comment (pelson) [inline lib/iris/aux_factory.py]: Can this be re-written as just:

```
[np.newaxis] * (cube.ndim - ndim) + [Ellipsis]
```
?

Comment (pelson) [inline lib/iris/aux_factory.py]: Actually, scrap that.

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1498,18 +1503,24 @@ def _sanitise_array(self, src, ndmin):
...
-        # desired (1,)
-        #   ... could change to: points = lazy.array(points, ndmin=1)
-        if not isinstance(points, iris.aux_factory.LazyArray):
+        # With the exception of LazyArrays, ensure points has an ndmin
+        # of 1 and is either a numpy or biggus array.
+        # This will avoid Scalar coords with points of shape () rather
+        # than the desired (1,)
+        if isinstance(points, biggus.Array):
+            if points.shape == ():
+                points = biggus.ArrayStack(np.array([points]))
```
Comment (pelson) [inline lib/iris/coords.py]: Could be a constant array.

**Comment by pelson** [comment]
:+1: from me. Over to @bjlittle.

P.S. I vote for removing LazyArray altogether for 2.0.

**Comment by rhattersley** [comment]
> P.S. I vote for removing LazyArray altogether for 2.0.

:+1: from me! I've been wanting to kill LazyArray for quite a while! :grin:

File: lib/iris/aux_factory.py [hunk 3]
```diff
@@ -384,6 +397,14 @@ def _shape(self, nd_values_by_key):
                     shape[i] = size
         return shape
 
+    def _dtype(self, arrays_by_key, **other_args):
+        dummy_args = {}
+        for key, array in arrays_by_key.iteritems():
+            dummy_args[key] = np.zeros(1, dtype=array.dtype)
+        dummy_args.update(other_args)
+        dummy_data = self._derive(**dummy_args)
```
Comment (bjlittle) [inline lib/iris/aux_factory.py]: Nice ...

File: lib/iris/coords.py [hunk 4]
```diff
@@ -446,8 +447,12 @@ def is_full_slice(s):
             points = self._points
             bounds = self._bounds
         else:
-            points = self.points
-            bounds = self.bounds
+            points = self._points
+            if isinstance(points, iris.aux_factory.LazyArray):
+                points = points.view()
```
Comment (bjlittle) [inline lib/iris/coords.py]: @rhattersley This forces the deferred coordinate data to actually load and cache, right?

If so, perhaps a comment here might be useful ...

File: lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb [hunk 5]
```diff
@@ -1257,26 +1250,33 @@ fc_extras
...
-            get_values = lambda: deferred_load(filename, nc_bounds_var_name)
-            bounds_data = iris.aux_factory.LazyArray(cf_bounds_var.shape,
-                                                     get_values)
+            dtype = cf_bounds_var.dtype
+            fill_value = getattr(cf_bounds_var.cf_data, '_FillValue',
+                                 netCDF4.default_fillvals[dtype.str[1:]])
+            proxy = iris.fileformats.netcdf.NetCDFDataProxy(
+                cf_bounds_var.shape, dtype, engine.filename,
+                cf_bounds_var.cf_name, fill_value)
+            bounds_data = biggus.OrthoArrayAdapter(proxy)
```
Comment (bjlittle) [inline lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb]: @rhattersley is it worthwhile having a private convenience function to generate the `biggus.OrthoArrayAdapter` for the `cf_coord_var` or the `cf_bounds_var` ... something along the lines of;

```python
from iris.fileformats.netcdf import NetCDFDataProxy as Proxy
...
def _to_biggus(engine, cf_var):
    dtype = cf_var.dtype
    fill_value = getattr(cf_var.cf_data, '_FillValue', netCDF4.default_fillvals[dtype.str[1:]])
    proxy = Proxy(cf_var.shape, dtype, engine.filename, cf_var.cf_name, fill_value)
    return biggus.OrthoArrayAdapter(proxy)
```

**Comment by bjlittle** [comment]
@rhattersley Golly, my head hurts :flushed: ... A couple of minor comments, otherwise awesome PR :+1:

**Comment by bjlittle** [comment]
> P.S. I vote for removing LazyArray altogether for 2.0.

:+1: ... and three votes makes it official!

File: lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb (hunk 5 above)
Comment (rhattersley) [inline lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb]: Nice idea.

File: lib/iris/coords.py (hunk 2 above)
Comment (rhattersley) [inline lib/iris/coords.py]: Nice idea - makes it clearer.

**Comment by rhattersley** [comment]
> @rhattersley Golly, my head hurts :flushed: ... A couple of minor comments, otherwise awesome PR :+1:

Sorry for the head trauma, but thanks for taking a look! I've added another commit to address the various comments.

**Comment by bjlittle** [comment]
@rhattersley Can't help yourself, huh? :smile: 

Once Travis is happy, so am I, thanks!

**Comment by rhattersley** [comment]
> I'm hoping this can be merged soon

Thanks to lots of work from lots of people ... it's in! :tada:

**Comment by ocefpaf** [comment]
Thanks @rhattersley for moving this forward!  Looking forward for 1.8.0!!

**Pull request merged**